File: packages/gbnf-experiment/data/20260909T005215Z_9a3e20fe/ported_implementation/test/buildRuleStack.test.ts
import assert from 'node:assert/strict';
import { describe, it } from 'node:test';

import {
  RuleChar,
  RuleCharExclude,
  RuleEnd,
} from '../src/grammarGraph/grammarGraphTypes.ts';
import type { Range, UnresolvedRule } from '../src/grammarGraph/grammarGraphTypes.ts';
import { RuleRef } from '../src/grammarGraph/ruleRef.ts';
import { buildRuleStack } from '../src/grammarParser/buildRuleStack.ts';
import {
  InternalRuleDefAlt,
  InternalRuleDefChar,
  InternalRuleDefCharAlt,
  InternalRuleDefCharNot,
  InternalRuleDefCharRngUpper,
  InternalRuleDefEnd,
  InternalRuleDefReference,
} from '../src/rulesBuilder/rulesBuilderTypes.ts';
import type { InternalRuleDef } from '../src/rulesBuilder/rulesBuilderTypes.ts';

const ichar = (value: number[]): InternalRuleDef => new InternalRuleDefChar(value);
const icharNot = (value: number[]): InternalRuleDef => new InternalRuleDefCharNot(value);
const icharAlt = (value: number): InternalRuleDef => new InternalRuleDefCharAlt(value);
const icharRngUpper = (value: number): InternalRuleDef => new InternalRuleDefCharRngUpper(value);
const iref = (value: number): InternalRuleDef => new InternalRuleDefReference(value);
const ialt = (): InternalRuleDef => new InternalRuleDefAlt();
const iend = (): InternalRuleDef => new InternalRuleDefEnd();

const char = (value: (number | Range)[]): UnresolvedRule => new RuleChar(value);
const charExclude = (value: (number | Range)[]): UnresolvedRule => new RuleCharExclude(value);
const ref = (value: number): UnresolvedRule => new RuleRef(value);
const end = (): UnresolvedRule => new RuleEnd();

interface Case {
  name: string;
  input: InternalRuleDef[];
  expected: UnresolvedRule[][];
}

const CASES: Case[] = [
  {
    name: "test_it_builds_rule_stack_for_a_single_path",
    input: [ichar([120])],
    expected: [[char([120]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_two_alternate_paths",
    input: [ichar([120]), ialt(), ichar([121])],
    expected: [[char([120]), end()], [char([121]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_three_alternate_paths",
    input: [ichar([120]), ialt(), ichar([121]), ialt(), ichar([122])],
    expected: [[char([120]), end()], [char([121]), end()], [char([122]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_char_not",
    input: [icharNot([120]), ialt(), icharNot([121]), ialt(), icharNot([122])],
    expected: [[charExclude([120]), end()], [charExclude([121]), end()], [charExclude([122]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_mixed_char_and_char_not",
    input: [icharNot([120]), ialt(), ichar([121]), ialt(), icharNot([122])],
    expected: [[charExclude([120]), end()], [char([121]), end()], [charExclude([122]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_char_not_with_two_characters_and_a_range",
    input: [icharNot([120]), icharAlt(121), icharAlt(122), icharRngUpper(130)],
    expected: [[charExclude([120, 121, [122, 130]]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_no_modifiers",
    input: [ichar([97]), icharRngUpper(122), iend()],
    expected: [[char([[97, 122]]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_no_modifiers",
    input: [ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), iend()],
    expected: [[char([[97, 122], [65, 90]]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_no_modifiers",
    input: [ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), icharAlt(48), icharRngUpper(57), iend()],
    expected: [[char([[97, 122], [65, 90], [48, 57]]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_question_mark_modifier",
    input: [ichar([97]), icharRngUpper(122), ialt(), iend()],
    expected: [[char([[97, 122]]), end()], [end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_question_mark_modifier",
    input: [ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), ialt(), iend()],
    expected: [[char([[97, 122], [65, 90]]), end()], [end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_question_mark_modifier",
    input: [ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), icharAlt(48), icharRngUpper(57), ialt(), iend()],
    expected: [[char([[97, 122], [65, 90], [48, 57]]), end()], [end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_plus_modifier",
    input: [ichar([97]), icharRngUpper(122), iref(1), ialt(), ichar([97]), icharRngUpper(122), iend()],
    expected: [[char([[97, 122]]), ref(1), end()], [char([[97, 122]]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_plus_modifier",
    input: [ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), iref(1), ialt(), ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), iend()],
    expected: [[char([[97, 122], [65, 90]]), ref(1), end()], [char([[97, 122], [65, 90]]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_plus_modifier",
    input: [ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), icharAlt(48), icharRngUpper(57), iref(1), ialt(), ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), icharAlt(48), icharRngUpper(57), iend()],
    expected: [[char([[97, 122], [65, 90], [48, 57]]), ref(1), end()], [char([[97, 122], [65, 90], [48, 57]]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_asterisk_modifier",
    input: [ichar([97]), icharRngUpper(122), iref(1), ialt(), iend()],
    expected: [[char([[97, 122]]), ref(1), end()], [end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_asterisk_modifier",
    input: [ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), iref(1), ialt(), iend()],
    expected: [[char([[97, 122], [65, 90]]), ref(1), end()], [end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_a_char_with_asterisk_modifier",
    input: [ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), icharAlt(48), icharRngUpper(57), iref(1), ialt(), iend()],
    expected: [[char([[97, 122], [65, 90], [48, 57]]), ref(1), end()], [end()]],
  },
  {
    name: "test_char_with_range",
    input: [ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), icharAlt(95), iend()],
    expected: [[char([[97, 122], [65, 90], 95]), end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_situation_root_ws_plus_newline_ws_star",
    input: [ichar([32]), icharAlt(92), icharAlt(110), iref(4), ialt(), iend()],
    expected: [[char([32, 92, 110]), ref(4), end()], [end()]],
  },
  {
    name: "test_it_builds_rule_stack_for_situation_root_char_plus_char_range",
    input: [ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), iref(1), ialt(), ichar([97]), icharRngUpper(122), icharAlt(65), icharRngUpper(90), iend()],
    expected: [[char([[97, 122], [65, 90]]), ref(1), end()], [char([[97, 122], [65, 90]]), end()]],
  },
];

describe('buildRuleStack', () => {
  const seen = new Map<string, number>();
  for (const { name, input, expected } of CASES) {
    const n = (seen.get(name) ?? 0) + 1;
    seen.set(name, n);
    const title = name.replace(/^test_/u, '').replace(/_/gu, ' ');
    it(`${title} (case ${n})`, () => {
      assert.deepStrictEqual(buildRuleStack(input), expected);
    });
  }

  it('throws when an alt appears before anything else', () => {
    assert.throws(() => buildRuleStack([ialt()]), /Encountered alt without anything before it/u);
  });

  it('throws when a char alt is not preceded by a char', () => {
    assert.throws(
      () => buildRuleStack([icharAlt(97)]),
      /Encountered char alt, should be handled by above block/u,
    );
  });

  it('throws on an unsupported rule type', () => {
    assert.throws(
      () => buildRuleStack([{ type: 'UNKNOWN' } as unknown as InternalRuleDef]),
      /Unsupported rule type/u,
    );
  });
});

File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/gbnf/grammar_parser/build_rule_stack_test.py
import pytest

from ..grammar_graph.grammar_graph_types import RuleChar, RuleCharExclude, RuleEnd
from ..grammar_graph.rule_ref import RuleRef
from ..rules_builder.rules_builder_types import (
    InternalRuleDefAlt,
    InternalRuleDefChar,
    InternalRuleDefCharAlt,
    InternalRuleDefCharNot,
    InternalRuleDefCharRngUpper,
    InternalRuleDefEnd,
    InternalRuleDefReference,
)
from .build_rule_stack import build_rule_stack


def ichar(value):
    return InternalRuleDefChar(value)


def ichar_alt(value):
    return InternalRuleDefCharAlt(value)


def ichar_rng_upper(value):
    return InternalRuleDefCharRngUpper(value)


def ichar_not(value):
    return InternalRuleDefCharNot(value)


def ialt():
    return InternalRuleDefAlt()


def iend():
    return InternalRuleDefEnd()


def iref(value):
    return InternalRuleDefReference(value)


def cp(char):
    return ord(char)


def make_range(lower, upper):
    return [
        lower if isinstance(lower, int) else cp(lower),
        upper if isinstance(upper, int) else cp(upper),
    ]


def test_it_builds_rule_stack_for_a_single_path():
    assert build_rule_stack([ichar([120])]) == [[RuleChar([120]), RuleEnd()]]


def test_it_builds_rule_stack_for_two_alternate_paths():
    assert build_rule_stack([ichar([cp("x")]), ialt(), ichar([cp("y")])]) == [
        [RuleChar([cp("x")]), RuleEnd()],
        [RuleChar([cp("y")]), RuleEnd()],
    ]


def test_it_builds_rule_stack_for_three_alternate_paths():
    assert build_rule_stack(
        [
            ichar([cp("x")]),
            ialt(),
            ichar([cp("y")]),
            ialt(),
            ichar([cp("z")]),
        ]
    ) == [
        [RuleChar([cp("x")]), RuleEnd()],
        [RuleChar([cp("y")]), RuleEnd()],
        [RuleChar([cp("z")]), RuleEnd()],
    ]


def test_it_builds_rule_stack_for_char_not():
    assert build_rule_stack(
        [
            ichar_not([cp("x")]),
            ialt(),
            ichar_not([cp("y")]),
            ialt(),
            ichar_not([cp("z")]),
        ]
    ) == [
        [RuleCharExclude([cp("x")]), RuleEnd()],
        [RuleCharExclude([cp("y")]), RuleEnd()],
        [RuleCharExclude([cp("z")]), RuleEnd()],
    ]


def test_it_builds_rule_stack_for_mixed_char_and_char_not():
    assert build_rule_stack(
        [
            ichar_not([cp("x")]),
            ialt(),
            ichar([cp("y")]),
            ialt(),
            ichar_not([cp("z")]),
        ]
    ) == [
        [RuleCharExclude([cp("x")]), RuleEnd()],
        [RuleChar([cp("y")]), RuleEnd()],
        [RuleCharExclude([cp("z")]), RuleEnd()],
    ]


def test_it_builds_rule_stack_for_char_not_with_two_characters_and_a_range():
    assert build_rule_stack(
        [
            ichar_not([cp("x")]),
            ichar_alt(cp("y")),
            ichar_alt(cp("z")),
            ichar_rng_upper(130),
        ]
    ) == [[RuleCharExclude([120, 121, make_range(122, 130)]), RuleEnd()]]


@pytest.mark.parametrize(
    ("grammar", "input", "expected"),
    [
        (
            "[a-z]",
            [ichar([cp("a")]), ichar_rng_upper(cp("z")), iend()],
            [[RuleChar([make_range("a", "z")]), RuleEnd()]],
        ),
        (
            "[a-zA-Z]",
            [
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                ichar_alt(cp("A")),
                ichar_rng_upper(cp("Z")),
                iend(),
            ],
            [[RuleChar([make_range("a", "z"), make_range("A", "Z")]), RuleEnd()]],
        ),
        (
            "[a-zA-Z0-9]",
            [
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                ichar_alt(cp("A")),
                ichar_rng_upper(cp("Z")),
                ichar_alt(cp("0")),
                ichar_rng_upper(cp("9")),
                iend(),
            ],
            [
                [
                    RuleChar(
                        [
                            make_range("a", "z"),
                            make_range("A", "Z"),
                            make_range("0", "9"),
                        ]
                    ),
                    RuleEnd(),
                ]
            ],
        ),
    ],
)
def test_it_builds_rule_stack_for_a_char_with_no_modifiers(grammar, input, expected):
    assert build_rule_stack(input) == expected


@pytest.mark.parametrize(
    ("grammar", "input", "expected"),
    [
        (
            "[a-z]?",
            [ichar([cp("a")]), ichar_rng_upper(cp("z")), ialt(), iend()],
            [
                [RuleChar([make_range("a", "z")]), RuleEnd()],
                [RuleEnd()],
            ],
        ),
        (
            "[a-zA-Z]?",
            [
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                ichar_alt(cp("A")),
                ichar_rng_upper(cp("Z")),
                ialt(),
                iend(),
            ],
            [
                [RuleChar([make_range("a", "z"), make_range("A", "Z")]), RuleEnd()],
                [RuleEnd()],
            ],
        ),
        (
            "[a-zA-Z0-9]?",
            [
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                ichar_alt(cp("A")),
                ichar_rng_upper(cp("Z")),
                ichar_alt(cp("0")),
                ichar_rng_upper(cp("9")),
                ialt(),
                iend(),
            ],
            [
                [
                    RuleChar(
                        [
                            make_range("a", "z"),
                            make_range("A", "Z"),
                            make_range("0", "9"),
                        ]
                    ),
                    RuleEnd(),
                ],
                [RuleEnd()],
            ],
        ),
    ],
)
def test_it_builds_rule_stack_for_a_char_with_question_mark_modifier(
    grammar, input, expected
):
    assert build_rule_stack(input) == expected


@pytest.mark.parametrize(
    ("grammar", "input", "expected"),
    [
        (
            "[a-z]+",
            [
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                iref(1),
                ialt(),
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                iend(),
            ],
            [
                [RuleChar([make_range("a", "z")]), RuleRef(1), RuleEnd()],
                [RuleChar([make_range("a", "z")]), RuleEnd()],
            ],
        ),
        (
            "[a-zA-Z]+",
            [
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                ichar_alt(cp("A")),
                ichar_rng_upper(cp("Z")),
                iref(1),
                ialt(),
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                ichar_alt(cp("A")),
                ichar_rng_upper(cp("Z")),
                iend(),
            ],
            [
                [
                    RuleChar([make_range("a", "z"), make_range("A", "Z")]),
                    RuleRef(1),
                    RuleEnd(),
                ],
                [RuleChar([make_range("a", "z"), make_range("A", "Z")]), RuleEnd()],
            ],
        ),
        (
            "[a-zA-Z0-9]+",
            [
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                ichar_alt(cp("A")),
                ichar_rng_upper(cp("Z")),
                ichar_alt(cp("0")),
                ichar_rng_upper(cp("9")),
                iref(1),
                ialt(),
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                ichar_alt(cp("A")),
                ichar_rng_upper(cp("Z")),
                ichar_alt(cp("0")),
                ichar_rng_upper(cp("9")),
                iend(),
            ],
            [
                [
                    RuleChar(
                        [
                            make_range("a", "z"),
                            make_range("A", "Z"),
                            make_range("0", "9"),
                        ]
                    ),
                    RuleRef(1),
                    RuleEnd(),
                ],
                [
                    RuleChar(
                        [
                            make_range("a", "z"),
                            make_range("A", "Z"),
                            make_range("0", "9"),
                        ]
                    ),
                    RuleEnd(),
                ],
            ],
        ),
    ],
)
def test_it_builds_rule_stack_for_a_char_with_plus_modifier(grammar, input, expected):
    assert build_rule_stack(input) == expected


@pytest.mark.parametrize(
    ("grammar", "input", "expected"),
    [
        (
            "[a-z]*",
            [ichar([cp("a")]), ichar_rng_upper(cp("z")), iref(1), ialt(), iend()],
            [
                [RuleChar([make_range("a", "z")]), RuleRef(1), RuleEnd()],
                [RuleEnd()],
            ],
        ),
        (
            "[a-zA-Z]*",
            [
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                ichar_alt(cp("A")),
                ichar_rng_upper(cp("Z")),
                iref(1),
                ialt(),
                iend(),
            ],
            [
                [
                    RuleChar([make_range("a", "z"), make_range("A", "Z")]),
                    RuleRef(1),
                    RuleEnd(),
                ],
                [RuleEnd()],
            ],
        ),
        (
            "[a-zA-Z0-9]*",
            [
                ichar([cp("a")]),
                ichar_rng_upper(cp("z")),
                ichar_alt(cp("A")),
                ichar_rng_upper(cp("Z")),
                ichar_alt(cp("0")),
                ichar_rng_upper(cp("9")),
                iref(1),
                ialt(),
                iend(),
            ],
            [
                [
                    RuleChar(
                        [
                            make_range("a", "z"),
                            make_range("A", "Z"),
                            make_range("0", "9"),
                        ]
                    ),
                    RuleRef(1),
                    RuleEnd(),
                ],
                [RuleEnd()],
            ],
        ),
    ],
)
def test_it_builds_rule_stack_for_a_char_with_asterisk_modifier(
    grammar, input, expected
):
    assert build_rule_stack(input) == expected


def test_char_with_range():
    input = [
        ichar([cp("a")]),
        ichar_rng_upper(cp("z")),
        ichar_alt(cp("A")),
        ichar_rng_upper(cp("Z")),
        ichar_alt(cp("_")),
        iend(),
    ]

    assert build_rule_stack(input) == [
        [
            RuleChar([make_range("a", "z"), make_range("A", "Z"), cp("_")]),
            RuleEnd(),
        ]
    ]


def test_it_builds_rule_stack_for_situation_root_ws_plus_newline_ws_star():
    input = [
        ichar([32]),
        ichar_alt(92),
        ichar_alt(110),
        iref(4),
        ialt(),
        iend(),
    ]

    assert build_rule_stack(input) == [
        [RuleChar([32, 92, 110]), RuleRef(4), RuleEnd()],
        [RuleEnd()],
    ]


def test_it_builds_rule_stack_for_situation_root_char_plus_char_range():
    input = [
        ichar([cp("a")]),
        ichar_rng_upper(cp("z")),
        ichar_alt(cp("A")),
        ichar_rng_upper(cp("Z")),
        iref(1),
        ialt(),
        ichar([cp("a")]),
        ichar_rng_upper(cp("z")),
        ichar_alt(cp("A")),
        ichar_rng_upper(cp("Z")),
        iend(),
    ]

    assert build_rule_stack(input) == [
        [
            RuleChar([make_range("a", "z"), make_range("A", "Z")]),
            RuleRef(1),
            RuleEnd(),
        ],
        [
            RuleChar([make_range("a", "z"), make_range("A", "Z")]),
            RuleEnd(),
        ],
    ]
